MiniDB/DataBase.cs

```diff
@@ -0,0 +1,1170 @@
+        /// <param name="notTransactionTypes">List of transaction types to skip</param>
+        /// <returns> the latest transaction that is not in notTransationTypes</returns>
+        private DBTransaction<T> GetLastTransaction(List<TransactionType> notTransactionTypes)
+        {
+            return this.ReleaseMutexOnError<DBTransaction<T>>(() =>
+            {
+                DBTransaction<T> last_transaction = null;
+                IEnumerable<DBTransaction<T>> temp_list = this.Transactions_DB;
+                do
+                {
+                    if (temp_list == null || temp_list.Count() == 0)
+                    {
+                        throw new DBCannotUndoException("Cannot find transaction");
+                    }
+
+                    var first_transaction = temp_list.FirstOrDefault().TransactionType;
+                    if (notTransactionTypes.Contains(first_transaction))
+                    {
+                        var count = this.CountRecentTransactions(notTransactionTypes, temp_list);
+                        temp_list = temp_list.Skip(count * 2);
+                    }
+                    else
+                    {
+                        last_transaction = temp_list.FirstOrDefault();
+                    }
+                }
+                while (last_transaction == null || notTransactionTypes.Contains(last_transaction.TransactionType));
+                return last_transaction;
+            });
+        }
+
+        /// <summary>
+        /// Invoked action and return its return value - if action throws an exception, release the mutex before re-raising
+        /// </summary>
+        /// <typeparam name="T2">The expected return type</typeparam>
+        /// <param name="action">The method to invoke</param>
+        /// <returns>The result of the action (if not error)</returns>
+        private T2 ReleaseMutexOnError<T2>(Func<T2> action)
+        {
+            try
+            {
+                return action.Invoke();
+            }
+            catch
+            {
+                this.Dispose();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Special case - templates can't handle a void type parameter --
+        /// also <see cref="ReleaseMutexOnError{T2}(Func{T2})" />
+        /// </summary>
+        /// <param name="action">the action to perform</param>
+        private void ReleaseMutexOnError(Action action)
+        {
+            try
+            {
+                action.Invoke();
+            }
+            catch
+            {
+                this.Dispose();
+                throw;
+            }
+        }
+        #endregion
+
+        /// <summary>
+        /// The information provided to a migration method
+        /// </summary>
+        public struct DBMigrationParameters
+        {
+            /// <summary>
+            /// The version of DB that is loaded
+            /// </summary>
+            public float OldVersion;
```

Comment:
Consider making these get-only and adding a constructor to set them, to better indicate that you don't expect these to change after construction.